### PR TITLE
Allow writing of experiment units table

### PR DESCRIPTION
**Pull request by lukesonnet**
## Features

This PR allows experimenters to write out a filtered and de-duped experiment units table to re-use across metrics rather than rescanning the full experiment assignment source.

* BigQuery users can now choose "pipeline mode" and write out an experiment units table to share across metrics tables, especially valuable if you have more than a couple metrics per experiment
* Activation metrics now respect the "Attribution model" of the experiment (e.g. you must activate within the First Exposure of your parent metric
* If your ratio metric has a non-binomial denominator, we now do not "cascade" your conversion windows and instead the numerator and denominator's own conversion windows and just aggregate all data within those windows from the exposure/activation timestamp.
* If your ratio metric has a binomial denominator, we keep the "cascading" conversion windows, but now they only kick off from that first experiment exposure/activation, they no longer will cascade from *any* experiment exposure, if you have the First Exposure attribution model selected

### Internal Changes

The **SQL integration** now has a couple of new methods and has a few changes:
* `getExperimentUnitsQuery` new - this creates the filtered, de-duped Experiment Units table, which can be used in a CTE if you aren't using the "Pipeline Mode", or can be used by `getExperimentUnitsTableQuery` to write the table back to the warehouse. The table defaults to expire in 24 hours
* `getExperimentMetricQuery` updated - this now either calls the created table or the CTE above, depending on whether we are in "Pipeline Mode", and then joins that table to the metrics queries

The **ExperimentResultsQueryRunner** now has dependencies between queries:
* Added the simple notion of `parentQueryIds` which will introduce some dependence between queries
* Added logic to wait for parent queries to finish before kicking of child query

The Data Source page has a whole new section to set up this writing functionality and I've edited the data model accordingly.

### SQL integration tester

* `base` experiments
  * All metrics are the same as before except for ratio metrics (expected)
  * All `pipeline` versions are the same as the non-pipeline versions for all metrics
  * Differences in binomial denominator ratio metrics are ONLY present when attribution model == "First Exposure" as expected.
* `activation` experiments
  * All different for "first exposure" settings because activation now has to happen in very first exposure
  * Results match again (except for ratio metrics) for "experiment duration" settings since activation can still happen from exp exposure until end of experiment, so this is reassuring that the changes in `activation` experiments are relegated to cases where the activation was outside that first experiment exposure window.

TODO:
- [X] Other experiments in sql integration tester that haven't been run yet (dimension, filter)
- [X] Test on other integrations (BigQuery, Snowflake, Postgres, clickhouse, MySQL tested so far)

### In app

Tested in app for reports and metrics, and results appear consistent, query modal looks reasonable (see video and screen shot below.

## Screenshots

Data source settings (only shows up in `type === "bigquery"` data sources):
Screenshot 2023-09-14 at 1 37 34 PM

Working in action:
https://www.loom.com/share/Big-Query-Data-Source-and-Experiment-Update-1682f46211b240af8caf2cc605484ee0

TODO:
- [X] Drop units table when all other queries finish
- [X] Allow user to customize whether to delete temp table
- [x] Enable only for BigQuery
- [x] Test write permissions when saving form WONTDO for V0
- [x] SQL integration tests
- [x] Test identity mappings explicitly
- [X] Add ability to track "query type" to compare performance WONTDO (can actually look at SQL if we need to)
- [ ] Docs around setting up dataset and when to use
- [x] Update Report runner as well

More notes in Notion as well: https://www.notion.so/growthbook/Tech-Spec-Data-Pipeline-Improvements-21ab7d6c104b4eee82d92cd3716180b0

Follow-ups?
* Store any identity join information in experiment units table so that we don't have to recreate it in the next step of the SQL
* Use results of units query for health page?
* Graphical view in browser of each step?

### CODE REVIEW

File: packages/front-end/components/Settings/EditDataSource/DataSourcePipeline/EditDataSourcePipeline.tsx [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    copy.settings.pipelineSettings = value;
+    await onSave(copy);
+  });
+
+  return (
+    <Modal
+      open={true}
+      submit={handleSubmit}
+      close={onCancel}
+      size="max"
```
Comment (jdorn) [inline packages/front-end/components/Settings/EditDataSource/DataSourcePipeline/EditDataSourcePipeline.tsx:41]: No need for this to be max size, it's just a few form fields

File: packages/front-end/components/Settings/EditDataSource/DataSourcePipeline/DataSourcePipeline.tsx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  children?: ReactNode;
+  titleClassName?: string;
+}> = ({ children, title = "", titleClassName = "", style }) => {
+  return (
+    <div className={`mb-2 ma-5 ${titleClassName}`} style={style}>
+      {title}
+      {children}
+    </div>
+  );
+};
```
Comment (jdorn) [inline packages/front-end/components/Settings/EditDataSource/DataSourcePipeline/DataSourcePipeline.tsx:22]: Not worth creating a component for this.  Only used 3 times and 1 of them is unique (bold text).  Better to just do it all inline

File: packages/front-end/components/Settings/EditDataSource/DataSourcePipeline/DataSourcePipeline.tsx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        In this section, you can configure how GrowthBook can use write
+        permissions to your Data Source to improve the performance of experiment
+        queries.
+      </div>
+      <div>
+        <DataSourcePipelineField
+          title="Pipeline Mode: "
+          titleClassName="font-weight-bold"
+        >
+          {dataSource.settings.pipelineSettings?.allowWriting
```
Comment (jdorn) [inline packages/front-end/components/Settings/EditDataSource/DataSourcePipeline/DataSourcePipeline.tsx:73]: You use `dataSource.settings.pipelineSettings` a bunch in this component.  Might be worth pulling this into a separate variable at the top.  Will make the rest easier to read

File: packages/front-end/components/Settings/EditDataSource/DataSourcePipeline/DataSourcePipeline.tsx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+          <>
+            <DataSourcePipelineField title="Destination dataset: ">
+              {dataSource.settings.pipelineSettings?.writeDataset ? (
+                <code>{dataSource.settings.pipelineSettings.writeDataset}</code>
+              ) : (
+                <em className="text-muted">not specified</em>
+              )}
+            </DataSourcePipelineField>
+            <DataSourcePipelineField title="Retention of temporary units table (hours): ">
+              {dataSource.settings.pipelineSettings?.unitsTableRetentionHours}
```
Comment (jdorn) [inline packages/front-end/components/Settings/EditDataSource/DataSourcePipeline/DataSourcePipeline.tsx:87]: The default is 24 hours if not specified, right?  So we should show that here if it's empty

File: packages/back-end/test/integrations/integration-query-generator.ts [hunk 5]
```diff
@@ -282,21 +301,25 @@ engines.forEach((engine) => {
...
       }
 
-      let activationMetrics: MetricInterface[] = [];
+      let activationMetric: MetricInterface | null = null;
       if (experiment.activationMetric) {
-        activationMetrics = allActivationMetrics.filter(
+        activationMetric = allActivationMetrics.filter(
           (m) => m.id === experiment.activationMetric
-        );
+        )[0];
```
[the remark this answers is not in the document]
Comment (lukesonnet) [reply packages/back-end/test/integrations/integration-query-generator.ts:308]: Ah yeah, this happened piecemeal so I didn't think to rewrite filter. Thanks.

File: packages/front-end/components/Settings/EditDataSource/DataSourcePipeline/DataSourcePipeline.tsx (hunk 2 above)
Comment (lukesonnet) [reply packages/front-end/components/Settings/EditDataSource/DataSourcePipeline/DataSourcePipeline.tsx:22]: Agreed. It did more in an earlier rendition.

File: packages/front-end/pages/datasources/[did].tsx [hunk 6]
```diff
@@ -345,6 +349,17 @@ mixpanel.init('YOUR PROJECT TOKEN', {
                   canEdit={canEdit}
                 />
               </div>
+
+              {d.type === "bigquery" && pipelineEnabled ? (
```
[the remark this answers is not in the document]
Comment (lukesonnet) [reply packages/front-end/pages/datasources/[did].tsx:353]: oof. i added that for a reason and then forgot about it. thanks!

**Review by jdorn — APPROVED** [review]

**Pull request merged**